HW2/P5/physics.pyx

```diff
@@ -55,32 +56,79 @@ cdef void sub_update(FLOAT[:, ::1] XY,
                      float R,
                      int i, int count,
                      UINT[:, ::1] Grid,
-                     float grid_spacing) nogil:
+                     float grid_spacing,
+                     uintptr_t locks_ptr,) nogil:
     cdef:
         FLOAT *XY1, *XY2, *V1, *V2
-        int j, dim
+        int j, dim, gx, gy, grid_size
         float eps = 1e-5
+        omp_lock_t *locks = <omp_lock_t *> <void *> locks_ptr
 
     # SUBPROBLEM 4: Add locking
-    XY1 = &(XY[i, 0])
-    V1 = &(V[i, 0])
+
+    # original code
+    #XY1 = &(XY[i, 0])
+    #V1 = &(V[i, 0])
+    # for j in range(i + 1, count):
+    #     XY2 = &(XY[j, 0])
+    #     V2 = &(V[j, 0])
+    #     if overlapping(XY1, XY2, R):
+    #         # SUBPROBLEM 4: Add locking
+    #         if not moving_apart(XY1, V1, XY2, V2):
+    #             collide(XY1, V1, XY2, V2)
+
+    #         # give a slight impulse to help separate them
+    #         for dim in range(2):
+    #             V2[dim] += eps * (XY2[dim] - XY1[dim])
+
     #############################################################
     # IMPORTANT: do not collide two balls twice.
     ############################################################
     # SUBPROBLEM 2: use the grid values to reduce the number of other
     # objects to check for collisions.
-    for j in range(i + 1, count):
-        XY2 = &(XY[j, 0])
-        V2 = &(V[j, 0])
-        if overlapping(XY1, XY2, R):
-            # SUBPROBLEM 4: Add locking
-            if not moving_apart(XY1, V1, XY2, V2):
-                collide(XY1, V1, XY2, V2)
-
-            # give a slight impulse to help separate them
-            for dim in range(2):
-                V2[dim] += eps * (XY2[dim] - XY1[dim])
 
+    # get grid index for current i
+    grid_size = int((1.0 / grid_spacing) + 1)
+
+    gx = min(grid_size - 1, max(0, <int>(XY[i, 0] / grid_spacing)))
+    gy = min(grid_size - 1, max(0, <int>(XY[i, 1] / grid_spacing)))
+
+    XY1 = &(XY[i, 0])
+    V1 = &(V[i, 0])
+
+    # update 5x5 neighborhood
+    for x in range(max(0, gx - 2), min(gx + 3, grid_size)):
+        for y in range(max(0, gy - 2), min(gy + 3, grid_size)):
+            # retrieve index of cell
+            j = Grid[x, y]
+
+            # only update if pair not visited yet
+            # (this here excludes also the case where gx == x, gy == y)
+            if j > i:
+                XY2 = &(XY[j, 0])
+                V2 = &(V[j, 0])
+                if overlapping(XY1, XY2, R):
+                    # SUBPROBLEM 4: Add locking
+                    if not moving_apart(XY1, V1, XY2, V2):
+
+                        # use locks for both i, j
+                        acquire(&locks[i])
```

Comment:
May be useful to acquire these locks prior to overlapping and moving_apart checks to prevent another thread from modifying in between.